### DISCUSSION TITLE
samtoolsのエラーについて

### DISCUSSION THREAD

**Discussion by hirokazu-okym** [nig-sc/nigsc_homepage2#422] [Q&A] [answered]
大変お世話になっております。
qsubに投げたsamtoolsが以下のエラーを出すようになり、
"error while loading shared libraries: libcrypto.so.10: cannot open shared object file: No such file or directory"
FAQの「何かプログラムを実行しようとすると、以下のようなライブラリに関するエラーが出ます。OS 移行前は問題なく実行できていました。」
に従い、.bash_profile, .bashrcを初期設定に戻し、samtoolsもインストール（コンパイル）し直しましたが、ゲートウェイノードでは実行できるのですが、
qloginすると実行できないようです。エラーの内容はほぼ同じで、
"samtools: error while loading shared libraries: libncursesw.so.5: cannot open shared object file: No such file or directory"
でした。
自分のログを見ると、1月15日の時点ではsamtoolsが正常に実行できていたようなので、もしかしたらOSの更新とは関係がないかもしれませんが、現状対処方法がわかりません。
対処法等をご教示いただけますと幸いです。
よろしくお願いいたします。

**Comment by sc-fs** [comment]
動作環境の確認させてください。samtoolsをコンパイルされているとのことですが、
こちらは、ゲートウェイノードで作成されたのでしょうか？
それとも、qlogin先になりますか？

また、1月15日の時点ではsamtoolsが正常に動作していたのは、ゲートウェイノードでしょうか？
それとも、qlogin先もしくは、ジョブ実行先の計算ノードでしょうか？

以上、よろしくお願いします。

**Reply by hirokazu-okym** [reply]
1月19日にsamtoolsをダウンロードし直して、コンパイルし直した際は、ゲートウェイノードだと思います。
OS更新前もゲートウェイノードでコンパイルしたと思います（2023年5月頃）。
1月15日は、qloginして（いると思う）qsubで実行しているので、ジョブ計算先だと思います（jobid? 24343571）。この時はsamtoolsが正常に実行されたみたいです。
1月19日に、15日と同じように（ひょっとしたらキューがmediumからshortに変わったかも）qsubで実行したら、
パイプラインが最後まで通らずに終了していて、エラーログファイルを見たら、最後に、
"samtools: error while loading shared libraries: libcrypto.so.10: cannot open shared object file: No such file or directory"
というメッセージが出ていて、ググってみたら、ライブラリの場所がみつけられない時のエラーということで、
確かに
$ ldd $(which samtools)
と打つと、ゲートウェイノードではすべてのライブラリの場所が表示されるのに、
ログインノードではいくつかのライブラリがみつけられないようでした。
さらにググっていたら、遺伝研スパコンのFAQにたどりつき、
.bash_profileと.bashrcを作り直して、samtools他のツールも全部コンパイルし直して、
再度
$ ldd $(which samtools)
とやってみたのですが、やはりログインノードでは見つからないライブラリが幾つかあるようです。

ただし、その後、もう一回qsubにパイプラインのスクリプトを投げてみたら、samtoolsは動いたように見えますが、その次のツールのところで止まってしまったみたいです。

素人なので状況説明が要領を得ていないかもしれませんが、よろしくお願い致します。

**Comment by sc-fs** [comment]
情報提供ありがとうございます。
osがUbuntu22.04に変更されたところでは、動作していないとのことですね。

samtoolsのソースセットにある、INSTALLを参照するとUbuntuではlibncurses5-dev5-devが必要とのことですが、
osがUbuntu22.04の計算ノードには、libncurses.so.5は無く、新しい、libncurses.so.6が存在しています。
実行時に発生していたエラーは、libncurses.so.5存在しないために、発生したことになります。

対応としては、新しいバージョンを古いバージョン名でも参照できるようにすることになります。
具体的には、
```
ln -s /usr/lib/x86_64-linux-gnu/libncursesw.so.6.3 /usr/lib/x86_64-linux-gnu/libncursesw.so.5
ln -s /usr/lib/x86_64-linux-gnu/libncurses.so.6.3 /usr/lib/x86_64-linux-gnu/libncurses.so.5
ln -s /usr/lib/x86_64-linux-gnu/libtinfo.so.6.3 /usr/lib/x86_64-linux-gnu/libtinfo.so.5

```
対応は、管理者権限が無いと対応できないので、こちらで対応しました。

お手数ですが、qlogin先もしくは,ジョブ投入(medium.q以外)でsamtoolsが動作出来るか確認をお願いします。

**Reply by hirokazu-okym** [reply]
有難うございます。
qloginしてエラーなくsamtoolsが実行できることを確認しました。
qsubの方はこの後確認してみます。
まずはお礼まで。

**Reply by hirokazu-okym** [reply]
連投で済みません。
qsubでもsamtoolsが動作するのを確認しました。
しかし、その次のarribaというツールが上手く動かなかったようです。

パイプラインでsamtoolsの座標ソートの後に、Bamファイルからキメラトランスクリプトの検出にarribaというツールを使っています。
念のため単独でarribaを動かすシェルスクリプトarriba.shを作って、ログインノード上で、
$ bash ./arriba.sh Aligned.out.csort.bam
とすると、ちゃんと動作するのを確認したのですが、このスクリプトをqsubで
qsub job_arriba.sh 
投げると、なぜか、
"./arriba.sh: line 11: arriba: command not found"
というメッセージが出て、動かないようです。
qsubのスクリプトは以下のように書いています。
job_arriba.sh
-----------------------------------------------------------------------------------------
#!/bin/bash

#$ -cwd 
# #$ -V 
#$ -l short 
#$ -l d_rt=24:00:00
#$ -l s_rt=24:00:00
# #$ -pe def_slot 4
#$ -l s_vmem=48G 
#$ -l mem_req=48G
#$ -N jobArriba
#$ -S /bin/bash

./arriba.sh Aligned.out.csort.bam
-----------------------------------------------------------------------------------------
何がまずいのでしょうか。

**Comment by sc-fs** [comment, marked as answer]
お手数をお掛けします。
PATHの設定が反映されていないため、エラーになっています。
お手数ですが、.bashrcの最後に設定されているexport及びalias設定を
.bash_profileの「source ~/.bashrc」行以降に次に同じexport及びalias設定を記載していただけないでしょうか？

**Reply by hirokazu-okym** [reply]
有難うございます。
.bash_profileにもパスの設定を書き込んだら、パイプラインが最後まで動くようになりました。
ただ、時々メモリ不足のためか、最後のarribaのところで、途中で終わってしまうことがあります。
もう少しqsubのメモリ確保を検討してみます。
